src/main/java/teammates/logic/api/EmailGenerator.java

```diff
@@ -379,18 +379,65 @@ private EmailWrapper generateSessionLinksRecoveryEmailForNonExistentStudent(Stri
+        if (linkFragmentsMap.isEmpty()) {
+            emailBody = Templates.populateTemplate(
+                    EmailTemplates.SESSION_LINKS_RECOVERY_ACCESS_LINKS_NONE,
+                    "${teammateHomePageLink}", Config.getFrontEndAppUrl("/").toAbsoluteString(),
+                    "${userEmail}", SanitizationHelper.sanitizeForHtml(recoveryEmailAddress),
+                    "${supportEmail}", Config.SUPPORT_EMAIL,
+                    "${sessionsRecoveryLink}", recoveryUrl);
+        } else {
+            var courseFragments = new StringBuilder(10000);
+            linkFragmentsMap.forEach((courseName, linksFragments) -> {
+                String courseBody = Templates.populateTemplate(
+                        EmailTemplates.FRAGMENT_SESSION_LINKS_RECOVERY_ACCESS_LINKS_BY_COURSE,
+                        "${sessionFragment}", linksFragments.toString(),
+                        "${courseName}", courseName);
+                courseFragments.append(courseBody);
+            });
+            emailBody = Templates.populateTemplate(
+                    EmailTemplates.SESSION_LINKS_RECOVERY_ACCESS_LINKS,
+                    "${userName}", SanitizationHelper.sanitizeForHtml(studentName),
+                    "${linksFragment}", courseFragments.toString(),
+                    "${userEmail}", SanitizationHelper.sanitizeForHtml(recoveryEmailAddress),
+                    "${teammateHomePageLink}", Config.getFrontEndAppUrl("/").toAbsoluteString(),
+                    "${supportEmail}", Config.SUPPORT_EMAIL,
+                    "${sessionsRecoveryLink}", recoveryUrl);
+        }
+
+        var email = getEmptyEmailAddressedToEmail(recoveryEmailAddress);
+        email.setType(EmailType.SESSION_LINKS_RECOVERY);
+        email.setSubjectFromType();
+        email.setContent(emailBody);
+        return email;
+    }
+
+    private Map<String, StringBuilder> generateLinkFragmentsMap(List<StudentAttributes> studentsForEmail) {
         var searchStartTime = TimeHelper.getInstantDaysOffsetBeforeNow(SESSION_LINK_RECOVERY_DURATION_IN_DAYS);
         Map<String, StringBuilder> linkFragmentsMap = new HashMap<>();
-        String studentName = null;
+        Map<String, String> courseIdToNameMap = new HashMap<>();
 
         for (var student : studentsForEmail) {
             RequestTracer.checkRemainingTime();
             // Query students' courses first
             // as a student will likely be in only a small number of courses.
             var course = coursesLogic.getCourse(student.getCourse());
             var courseId = course.getId();
+            String courseName;
+
+            if (courseIdToNameMap.containsKey(courseId)) {
+                courseName = courseIdToNameMap.get(courseId);
+            } else {
+                courseName = coursesLogic.getCourse(courseId).getName();
+                courseIdToNameMap.put(courseId, courseName);
+            }
```

Comment:
Is it not possible to just use `Course::getName` for `courseName`? Otherwise, this adds an extra `getCourse` call which does not seem necessary (unless I'm missing something here).

```suggestion
            var course = coursesLogic.getCourse(student.getCourse());
            var courseId = course.getId();
            String courseName = course.getName();
```

Reply:
Hi Jay, nice catch! It was carried over from the older implementation, I thought it was a datastore thing, but I looked into it and it seems to work! Have made the changes

Reply:
Hi Jay, might have to revert this change, I'm not sure if course name is unique and if it isn't not sure if it will affect the correctness. Will look into this later

Reply:
Oh yea, I think you're right. The course name is not unique. In that case, maybe we can consider using `CourseAttributes` as the key of the `linkFragmentsMap`? This can help keep your optimisation because you will still be able to avoid the call to `getCourse`.

Reply:
Yup this works! Have made the change